### PR TITLE
macros: improve `Span`'s expansion information

### LINKED ISSUE #39450: Spans for Paths can be incorrect
Consider the following example (from term-painter):

```
macro_rules! impl_format {
    ($symbol:expr, $fmt:ident) => {
        impl<T: fmt::$fmt> fmt::$fmt for Painted<T> {
            fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
                panic!()
            }
        }
    }
}

impl_format!("{}", Display);
```

On the third line are two paths of the form `fmt::$fmt`, after expansion, these are both `fmt::Display`. The way spans for paths are calculated we take the start of `fmt` and the end of `Display` (see `parse_path` in parser.rs), this means the span for the path is 9 lines long and pretty incoherent. Annoyingly, this span does not get an expansion id, and thus has no indication that it came from generated code. This is visible in error messages:

```
   |
9  |           impl<T: fmt::$fmt> fmt::$fmt for Painted<T> {
   |  _________________^ starting here...
10 | |             fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
11 | |                 panic!()
12 | |             }
13 | |         }
14 | |     }
15 | | }
16 | |
17 | | impl_format!("{}", Display);
   | | -------------------------^--
   | |_|________________________|
   |   |                        ...ending here
   |   in this macro invocation
```

### LINKED ISSUE #40649: macros: The line number that `panic!` reports can be wrong in macro invocations
For [example](https://play.rust-lang.org/?gist=e3c24e819b9250e11399c19c682c8b2d&version=nightly&backtrace=0),
```rust
macro_rules! identity_macro { ($($t:tt)*) => { $($t)* } }

fn main() {
    identity_macro! { //< Today, the message says we panic here.
        panic!("")
    }
}
```

### PR DESCRIPTION
**Pull request by jseyfried**
This PR improves `Span`'s expansion information. More specifically:
 - It refactors AST node span construction to preserve expansion information.
   - Today, we only use the underlying tokens' `BytePos`s, throwing away the `ExpnId`s.
   - This improves the accuracy of AST nodes' expansion information, fixing #30506.
 - It refactors `span.expn_id: ExpnId` to `span.ctxt: SyntaxContext` and removes `ExpnId`.
   - This gives all tokens as much hygiene information as `Ident`s.
   - This is groundwork for procedural macros 2.0 `TokenStream` API.
   - This is also groundwork for declarative macros 2.0, which will need this hygiene information for some non-`Ident` tokens.
 - It simplifies processing of spans' expansion information throughout the compiler.
 - It fixes #40649.
 - It fixes #39450 and fixes part of #23480.

r? @nrc

### CODE REVIEW

**Comment by jseyfried** [comment]
cc @eddyb 
cc #38356 #39412

File: src/test/run-pass/syntax-extension-source-utils.rs [hunk 1]
```diff
@@ -22,7 +22,7 @@ macro_rules! indirect_line { () => ( line!() ) }
 
 pub fn main() {
     assert_eq!(line!(), 24);
-    assert_eq!(column!(), 4);
+    assert_eq!(column!(), 15);
```
Comment (jseyfried) [inline src/test/run-pass/syntax-extension-source-utils.rs]: c.f. #40649; technically a [breaking-change]

**Review by nrc — APPROVED** [review]
r=me, assuming losing the command line span thing is fine.

This seems a good step in the right direction. I worry a bit that `to` is still too easy to create malformed spans, but it is def. better than mk_sp.

File: src/libsyntax_pos/lib.rs [hunk 2]
```diff
@@ -157,12 +224,7 @@ impl fmt::Debug for Span {
     }
 }
 
-pub const DUMMY_SP: Span = Span { lo: BytePos(0), hi: BytePos(0), expn_id: NO_EXPANSION };
-
-// Generic span to be used for code originating from the command line
-pub const COMMAND_LINE_SP: Span = Span { lo: BytePos(0),
-                                         hi: BytePos(0),
-                                         expn_id: COMMAND_LINE_EXPN };
```
Comment (jseyfried) [inline src/libsyntax_pos/lib.rs]: Same as before -- `DUMMY_SP`. The only place we used `COMMAND_LINE_SP` was used was [here](https://github.com/rust-lang/rust/pull/40597/files#diff-0d146cdcf8b61538ee2f552c43719757L84), which only happens when naming a legacy plugin crate through a debugging command line option.

**Comment by arielb1** [comment]
-  #40821
    @bors retry

**Comment by arielb1** [comment]
-  #40821
    @bors retry

**Comment by alexcrichton** [comment]
Hm that error may not be spurious?

**Comment by frewsxcv** [comment]
I don't think try works

@bors try-

**Comment by SergioBenitez** [comment]
I'm seeing the following, which may be related to this:

```rust
warning: {warning message from a syntax extension}
--> /a/nice/big/file/path/here.rs:10:21
 |
 | and so on...
 |
 = note: this error originates in a macro outside of the current crate
```

The issue is that the "note:` refers to the warning as an "error". It should say: "this warning originates from a macro outside of the current crate".

**Comment by jseyfried** [comment]
@Sergio This should be fixed by #40939 (as discussed on IRC).

**Pull request merged**